### PR TITLE
Add Deep CI chunk fallback and harden arbiter backlog contract

### PR DESCRIPTION
**Pull request by KjellKod**
## Summary
Adds Deep CI oversized-file chunk fallback and hardens the plan-phase arbiter backlog contract against schema drift so canonical review artifacts are validated and atomically published by the orchestrator, not hand-written by a model.
- Track A: selected files larger than `DEEP_CI_MAX_FILE_CHARS` now render deterministic changed-line context chunks with explicit `Mode:` metadata instead of being silently skipped.
- Track B: arbiter emits only `arbiter_verdict.md` + canonical `review_findings.json`; orchestrator validates findings, builds backlog, and atomically publishes from `.next` scratch files; retries arbiter exactly once on validator failure.

## Changes

- **Deep CI (.github/)**:
  - New constants in `.github/scripts/codex_review.py`: `DEEP_CI_CHUNK_CONTEXT_LINES=100`, `DEEP_CI_MAX_CHUNKS_PER_FILE=4`, `DEEP_CI_MAX_CHUNK_CHARS=12000`, `DEEP_CI_MAX_FETCH_CHARS=200000`.
  - New helpers: `parse_changed_line_ranges`, `build_line_windows`, `extract_line_chunk`, `fit_chunk_to_char_cap`.
  - `fetch_deep_ci_files()` threads parsed diff ranges from `/tmp/pr.diff`; `render_deep_ci_context()` emits `Mode: full-file | chunked-large-file | skipped` with reasons (`hard-cap`, `no-range`, `total-cap-exhausted`).
  - Chunked rendering reports `changed_lines_included` vs `changed_lines_omitted` under per-chunk cap pressure.
  - Parser handles paths with spaces, rename/new-path diffs, omitted-count hunk headers (`@@ -a +b @@`), and no longer misclassifies in-hunk `+++ ` added-line content as a file header.
  - Prompt text in `.github/codex-review-prompt.md` distinguishes full vs chunked partial context and preserves the RIGHT-side-only Deep CI findings constraint.

- **Arbiter contract (.skills/quest/)**:
  - `.skills/quest/agents/arbiter.md`: drops backlog authorship from the output contract; inlines `REQUIRED_FINDING_FIELDS`, allowed `severity`/`confidence` enums, and a minimal valid finding example.
  - `.skills/quest/delegation/workflow.md` Step 3 (Invoke Arbiter): clean stale `.next`, run `validate-findings`, retry arbiter once on failure, run `build-backlog --phase plan`, run `validate-backlog`, atomic `os.replace` publish, stop route on any failure, iterate/`solo`/stale-scratch handling, rollout-ordering constraint.

- **Orchestration (scripts/)**:
  - `scripts/quest_review_intelligence.py`: `build-backlog --phase {plan,review}`; new `validate-backlog` CLI; plan-phase defaults (`decision=fix_now`, `owner=builder`, `batch=<kind>-<path_group>`).
  - `scripts/quest_runtime/review_intelligence.py`: `_path_group_from_finding` uses first directory segment with `root` fallback; decoupled plan-phase vs review-phase decision policies; runtime-exposed `validate_review_backlog`.
  - `scripts/quest_runtime/artifacts.py`: arbiter `ROLE_ARTIFACTS` retargeted to `review_findings.json.next` so canonical findings are never zeroed by pre-run truncation.
  - `scripts/quest_claude_runner.py`: opt-in JSONL telemetry seam gated by `QUEST_RUNNER_TELEMETRY_LOG`; emits `attempt_start` / `attempt_end` events on every invocation; zero production overhead when unset.
  - `scripts/quest_validate-quest-state.sh`: accepts `--quest-dir`/`--target` flags and supports checkpoint re-validation so the explicit validator command works after the quest has advanced phases.

- **Tests**:
  - `tests/unit/test_codex_review.py`: new coverage for F1 hard-cap `Mode: skipped`, F2 chunk-metadata honesty under cap, F3 path-with-spaces / rename / new-path diffs, F4 omitted-count hunk header, in-hunk `+++ ` content, total-cap-exhausted render, and unchanged full-file behavior.
  - `tests/unit/test_review_intelligence.py`: plan-phase defaults, root-batch fallback, review-phase policy preservation.
  - `tests/unit/test_artifacts.py`: arbiter findings `.next` retarget.
  - `tests/test-quest-runtime.sh`: helper-level retry harness plus runner-path test that invokes `scripts/quest_claude_runner.py` with a fake bridge, parses `QUEST_RUNNER_TELEMETRY_LOG`, and asserts `attempts==2`, `retries==1`, no `transition` event before `publish`, and canonical `(mtime_ns, inode, size, sha256)` snapshots unchanged until atomic publish.

## Validation

- [x] **Unit tests — Deep CI**: Action: `python3 -m pytest tests/unit/test_codex_review.py -x`. Expected: 70 passed, including `test_deep_ci_hard_cap_file_renders_mode_skipped_with_reason`, `test_chunk_metadata_reports_included_and_omitted_changed_lines_under_cap_pressure`, `test_parse_changed_line_ranges_handles_paths_with_spaces_and_rename_new_paths`, `test_parse_changed_line_ranges_handles_omitted_count_hunk_header`, and `test_parse_changed_line_ranges_does_not_treat_in_hunk_plus_plus_space_as_header`.

- [x] **Unit tests — review intelligence**: Action: `python3 -m pytest tests/unit/test_review_intelligence.py -x`. Expected: 27 passed, including `test_build_backlog_plan_phase_defaults_to_fix_now_builder_and_deterministic_batch_slug`, `test_build_backlog_plan_phase_uses_root_batch_for_root_level_paths`, and `test_build_backlog_preserves_review_phase_policy_for_code_review_findings`.

- [x] **Full Python suite**: Action: `python3 -m pytest tests/ -x`. Expected: 445 passed.

- [x] **Quest runtime harness**: Action: `bash tests/test-quest-runtime.sh`. Expected: 26 passed / 0 failed. Both `test_plan_review_retry_harness_preserves_canonical_artifacts_until_publish` (helper-level) and `test_plan_review_retry_via_runner_preserves_canonical_artifacts_until_publish` (runner-path with stat-snapshot invariants) pass.

- [x] **Validate-quest-state harness**: Action: `bash tests/test-validate-quest-state.sh`. Expected: 47 passed / 0 failed. New `--quest-dir/--target` flag parsing and plan-phase checkpoint re-validation exercised.

- [ ] **Walkthrough — exercise the arbiter retry contract manually**: Prerequisites: working tree on this branch, Python 3 available. Action: read `.skills/quest/delegation/workflow.md` around "Invoke Arbiter" section and trace through the validate → retry → build-backlog → atomic publish flow; then inspect `scripts/quest_runtime/artifacts.py` to confirm arbiter's findings target is `review_findings.json.next`. Expected: single validator retry cap is documented; stop route on failure does not call `quest_state.py --transition plan_reviewed`; canonical `review_findings.json` / `review_backlog.json` are only written by `os.replace` from `.next`.

Watch for: the `QUEST_RUNNER_TELEMETRY_LOG` seam is opt-in; if an unrelated test ever sets this env var, runner telemetry will append there.

## Notes

- Rollout ordering inside this PR follows the `arb-it2-b4` constraint: workflow/helper/runtime edits land before the arbiter contract trim in `arbiter.md`, so no intermediate state breaks the `plan_reviewed` transition. All changes ship in one commit so this is a deploy-ordering invariant for reviewers to validate, not a merge-ordering one.
- The builder also extended `scripts/quest_validate-quest-state.sh` with `--quest-dir`/`--target` flags. The positional form used by existing callers still works; `tests/test-validate-quest-state.sh` covers both.
- Two fix iterations were needed: iteration 1 resolved a parser `+++ ` edge case and a batch-slug root-fallback mismatch; iteration 2 strengthened the B5 runner-path test to assert transition ordering from runtime telemetry and canonical immutability from `(mtime_ns, inode, size, sha256)` snapshots rather than from a local counter and content-equality.

<pre>
     ▐▛███▜▌
    ▝▜█████▛▘
      ▘▘ ▝▝
Quest/Co-Authored by
Co-Authored-By: Claude Opus 4.7 <noreply@anthropic.com>
Co-Authored-By: Codex <noreply@openai.com>
in collaboration with KjellKod <kjell.hedstrom@gmail.com>
</pre>

### CODE REVIEW

File: .github/scripts/codex_review.py [hunk 1]
```diff
@@ -191,52 +493,187 @@ def fetch_deep_ci_files(
...
+                    f"full file exceeded {max_chars_per_file} chars; only changed-line windows are included"
+                ),
+            }
+        )
 
     return snapshots
 
 
 def render_deep_ci_context(snapshots, selected_files=None):
     """Render Deep CI snapshots and omission notes as prompt markdown."""
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .github/scripts/codex_review.py:631]: Good catch — same finding surfaced again at line 660 after the follow-up commit. Fixed in f010b23: the `Changed RIGHT-side lines in this chunk` line now reads `changed_lines_included` (falling back to `changed_lines` when absent) so the label matches the content actually in the snippet. The separate `changed_lines_omitted` block continues to surface trimmed lines explicitly.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

File: .github/scripts/codex_review.py [hunk 2]
```diff
@@ -191,52 +493,187 @@ def fetch_deep_ci_files(
...
+            f"Included line ranges: {ranges}",
+            f"Omitted: {snapshot.get('reason', '')}",
+        ]
+        for chunk in chunks:
+            chunk_header = f"### {path} lines {chunk['start_line']}-{chunk['end_line']}"
+            chunk_lines = [
+                chunk_header,
+                (
+                    "Changed RIGHT-side lines in this chunk: "
+                    + ", ".join(str(line) for line in chunk.get("changed_lines", []))
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .github/scripts/codex_review.py]: Agreed — label was dishonest under cap pressure, undermining the F2 invariant. Fixed in f010b23:

```python
"Changed RIGHT-side lines in this chunk: "
+ ", ".join(
    str(line)
    for line in chunk.get(
        "changed_lines_included", chunk.get("changed_lines", [])
    )
)
```

The `changed_lines_omitted` block below remains as the explicit omission marker. All 70 `test_codex_review.py` tests green after the change.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

File: scripts/quest_claude_runner.py [hunk 3]
```diff
@@ -5,11 +5,43 @@
...
+    if not log_path:
+        return
+    try:
+        path = Path(log_path)
+        path.parent.mkdir(parents=True, exist_ok=True)
+        record = dict(event)
+        record.setdefault("ts", time.time())
+        with path.open("a", encoding="utf-8") as handle:
+            handle.write(json.dumps(record, ensure_ascii=True) + "\n")
+    except OSError:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_claude_runner.py]: Fair point — the docstring's "must never affect the runner outcome" contract was stronger than `except OSError` could guarantee. Fixed in f010b23:

```python
except Exception:
    # Telemetry is best-effort; never fail the runner because of it.
    # Broad except is intentional — the docstring contract is that
    # telemetry must never affect runner outcome, which includes
    # TypeError from json.dumps on unexpected payloads.
    return
```

Broad-except is now explicit-and-commented so future readers understand it's deliberate, not lazy. Runner tests remain green.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

File: .skills/quest/delegation/workflow.md [hunk 4]
```diff
@@ -494,27 +498,49 @@ gates.max_plan_iterations (default: 4)
...
+     - Publish atomically only after both files validate:
+       - `os.replace(".quest/<id>/phase_01_plan/review_findings.json.next", ".quest/<id>/phase_01_plan/review_findings.json")`
+       - `os.replace(".quest/<id>/phase_01_plan/review_backlog.json.next", ".quest/<id>/phase_01_plan/review_backlog.json")`
+     - On publish failure: STOP route, log full error, do not transition, and leave `.next` files in place.
+
+6. **Check verdict and transition guard:**
    - If `NEXT: builder`:
-     - Plan approved! Transition state atomically: `python3 scripts/quest_state.py --quest-dir .quest/<id> --transition plan_reviewed --status complete --last-verdict approve --expect-phase plan` — if this fails, report the validation error to the user and STOP. Do NOT modify state.json manually. Then proceed to **Step 3.5** (Interactive Presentation). Do not attempt the `presenting` transition while state still says `phase: plan`.
+     - Transition only after successful `validate-findings` + `build-backlog --phase plan` (+ optional `validate-backlog`) + atomic publish.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/quest/delegation/workflow.md]: Good callout — the guard sentence sat inside the workflow branch but read universally, which would confuse anyone implementing solo support. Fixed in f010b23 with an explicit carve-out:

> **Workflow mode only:** Transition only after successful `validate-findings` + `build-backlog --phase plan` (+ optional `validate-backlog`) + atomic publish. In solo mode the arbiter is skipped (see item 5 above), and so is the validate/build/publish pipeline — transition directly without those prerequisites.

This matches the existing solo-mode branch at line 471-476 ("Skip Arbiter entirely … Do not require review_backlog.json"). No test changes needed since this is a docs-only clarification.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

File: ideas/2026-04-22-review-ergonomics-and-team-preference-memory.md [hunk 5]
```diff
@@ -0,0 +1,378 @@
+---
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply ideas/2026-04-22-review-ergonomics-and-team-preference-memory.md:1]: Noted — and a fair scope observation. To be transparent, `ideas/2026-04-22-review-ergonomics-and-team-preference-memory.md` was added in commit 291312d independently of the Deep CI / arbiter-contract work (both PR-visible commits 54b1d4e and 1e68193 are cleanly scoped to those two tracks).

@KjellKod — your call on whether to split this out into its own PR or keep it bundled. If you'd like me to revert the `ideas/` commit on this branch and open a separate PR for it, I'm happy to; otherwise leaving it as-is.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

File: scripts/quest_runtime/review_intelligence.py [hunk 6]
```diff
@@ -309,12 +371,71 @@ def build_review_backlog(
...
         "generated_at": utc_now_iso(),
         "at_loop_cap": at_loop_cap,
+        "phase": phase,
         "allowed_decisions": list(ALLOWED_DECISIONS),
         "counts": counts,
         "items": items,
     }
 
 
+def validate_review_backlog(backlog: Any) -> list[str]:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/review_intelligence.py:381]: Fixed in 7649494. `validate_review_backlog()` now requires a top-level `phase` and rejects values outside `plan|review`; `--expected-phase` remains the stricter equality check. Added coverage for missing and invalid phase in `tests/unit/test_review_intelligence.py`.

- Reviewed by Codex, in collaboration with @KjellKod

File: .github/scripts/codex_review.py [hunk 7]
```diff
@@ -191,52 +493,192 @@ def fetch_deep_ci_files(
...
+            fitted_chunk = fit_chunk_to_char_cap(
+                raw_chunk,
+                changed_lines,
+                cap=max_chunk_chars,
+            )
+            chunk_content = fitted_chunk["content"]
+            if not chunk_content:
+                continue
+            chunk_chars = len(chunk_content)
+            if total_chars + chunk_chars > max_total_chars:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .github/scripts/codex_review.py:584]: Fixed in 7649494. When the Deep CI total budget is exhausted mid-file, the fetcher records the remaining planned windows in `total_cap_omitted_windows`, and the renderer prints a `Total-cap omitted` line with the dropped ranges/count. Added renderer coverage in `tests/unit/test_codex_review.py`.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_runtime/review_intelligence.py [hunk 8]
```diff
@@ -309,12 +371,83 @@ def build_review_backlog(
...
+    # Baseline phase enforcement: the top-level 'phase' field must be present
+    # and one of the allowed values. --expected-phase is the stricter equality
+    # check layered on top; this is the catch-all for typos/corruption/missing.
+    phase = backlog.get("phase")
+    if "phase" not in backlog:
+        errors.append("backlog must include a top-level 'phase' field")
+    elif not isinstance(phase, str) or phase not in ("plan", "review"):
+        errors.append(
+            f"backlog 'phase' must be one of 'plan' or 'review', got {phase!r}"
+        )
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/review_intelligence.py]: Fixed in 6929c53. The validator now checks against `ALLOWED_BACKLOG_PHASES`, and the regression assertion derives its expected allowed-values string from that same constant so future phase additions do not drift.

- Reviewed by Codex, in collaboration with @KjellKod

File: .claude/hooks/enforce-allowlist.sh [hunk 9]
```diff
@@ -60,6 +58,39 @@ check_file_write() {
...
+  local regex="^"
+  local i char next
+
+  for ((i = 0; i < ${#pattern}; i++)); do
+    char="${pattern:i:1}"
+
+    if [[ "$char" == "*" ]]; then
+      next="${pattern:i+1:1}"
+      if [[ "$next" == "*" ]]; then
+        regex+=".*"
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .claude/hooks/enforce-allowlist.sh]: Fixed in 4873a95. `check_file_write()` now normalizes the requested write path against `REPO_ROOT` before glob matching and rejects paths that resolve outside the repo. Added an integration regression showing an `arbiter_agent` write through `.quest/../scripts/quest_state.py` is blocked while normal `.quest/**` writes still pass.

- Reviewed by Codex, in collaboration with @KjellKod

File: .github/scripts/codex_review.py [hunk 10]
```diff
@@ -161,20 +167,316 @@ def markdown_code_fence(content):
...
+            )
+        else:
+            merged.append(window)
+
+    if len(merged) > max_chunks:
+        ranked = sorted(
+            merged,
+            key=lambda item: (
+                -item["changed_line_count"],
+                item["start_line"],
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .github/scripts/codex_review.py:313]: Fixed in 4873a95. `build_line_windows(..., include_omitted=True)` now returns windows dropped by the per-file chunk cap, `fetch_deep_ci_files()` carries them as `chunk_cap_omitted_windows`, and `render_deep_ci_context()` prints a `Chunk-cap omitted` line with the omitted ranges/count. Added unit coverage for the window plan and rendered output.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_validate-quest-state.sh [hunk 11]
```diff
@@ -368,6 +368,50 @@ $item_errors"
...
+    fail "Semantic check: plan-phase backlog check failed ($backlog_file): $phase_output"
+    return 1
+  fi
+
+  local bad_decisions
+  bad_decisions=$(jq -r '
+    [
+      (.items // []) | to_entries[] |
+      select((.value.decision // "") as $d |
+        ($d != "fix_now") and ($d != "verify_first")
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_validate-quest-state.sh]: Agreed — the canonical `_plan_phase_decision` in `scripts/quest_runtime/review_intelligence.py` hardcodes `decision: "fix_now"`, so `verify_first` under `phase: "plan"` is contract drift by definition. Previously the validator allowed both on pragmatic grounds (both flow to the builder), but that tolerance undermines Track B's whole "validator enforces the canonical contract" point.

Fixed in 7545f5b. The jq guard in `validate_plan_phase_backlog` now requires every item to be exactly `fix_now`:

```bash
select((.value.decision // "") != "fix_now") |
"[" + (.key | tostring) + "] non-canonical plan-phase decision: " + (.value.decision // "missing")
```

Error message changed from "non-actionable" to "non-canonical" to reflect the tighter contract. Added `test_plan_to_plan_reviewed_rejects_verify_first_decision` and updated the existing non-actionable test. review-phase backlog semantics are unaffected. All gates green: 455 pytest, 26 quest-runtime, 50 validate-quest-state.

- Reviewed by Claude Opus 4.7, in collaboration with @KjellKod

[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_validate-quest-state.sh]: Follow-up pushed in 9c44a60. The production guard in 7545f5b already checks the full canonical plan defaults: decision=fix_now, owner=builder, and the deterministic <kind>-<path-group> batch slug. 9c44a60 adds regression coverage that a phase=plan backlog with review-style owner/batch defaults is rejected, alongside the verify_first regression.

Validated locally with:
- bash tests/test-validate-quest-state.sh
- python3 -m pytest tests/unit/test_review_intelligence.py -q
- bash scripts/quest_validate-manifest.sh
- git diff --check -- scripts/quest_validate-quest-state.sh tests/test-validate-quest-state.sh

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_validate-quest-state.sh [hunk 12]
```diff
@@ -368,6 +368,100 @@ $item_errors"
...
+    def slugify($fallback):
+      ascii_downcase
+      | gsub("[^a-z0-9]+"; "-")
+      | gsub("^-+"; "")
+      | gsub("-+$"; "") as $slug
+      | if $slug == "" then $fallback else $slug end;
+
+    def path_group:
+      (if (.write_scope | type) == "array" then
+        ([.write_scope[] | select(type == "string") | trim_string | select(. != "")] | sort | first // "")
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_validate-quest-state.sh]: Fixed in a028bb9. cubic found that the shell validator trimmed write_scope entries before sorting, while the canonical Python backlog builder sorts the original strings and trims only after selecting the candidate. The jq path_group now preserves raw ordering while still filtering whitespace-only entries.

Added test_plan_to_plan_reviewed_accepts_canonical_raw_write_scope_sorting to cover the false validation failure. Validated with bash tests/test-validate-quest-state.sh, python3 -m pytest tests/unit/test_review_intelligence.py tests/unit/test_pr_review_cycle.py -q, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: .skills/quest/delegation/workflow.md [hunk 13]
```diff
@@ -494,27 +498,49 @@ gates.max_plan_iterations (default: 4)
...
+       - Keep canonical artifacts untouched; keep `.next` files for debugging.
+   - If arbiter handoff says `next: planner`:
+     - Skip `build-backlog` and skip atomic publish.
+     - Clean `.next` scratch files.
+     - Preserve canonical `review_findings.json` / `review_backlog.json`.
+     - Return control to planner refinement loop.
+   - If arbiter handoff says `next: builder` and findings validation passed:
+     - Build backlog from validated findings:
+       - `python3 scripts/quest_review_intelligence.py build-backlog --phase plan --findings .quest/<id>/phase_01_plan/review_findings.json.next --output .quest/<id>/phase_01_plan/review_backlog.json.next`
+     - Optional guard:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/quest/delegation/workflow.md]: Fixed in 2e6019c. The workflow now makes the pre-publish backlog validation mandatory and calls validate-backlog with --expected-phase plan before atomic os.replace, so a review-phase backlog cannot be published as the canonical plan artifact.

Validated locally with:
- bash scripts/quest_validate-manifest.sh
- git diff --check -- .skills/quest/delegation/workflow.md scripts/quest_validate-quest-state.sh tests/test-validate-quest-state.sh

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_validate-quest-state.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_validate-quest-state.sh]: Fixed in a028bb9. The jq mirror now filters blank write_scope entries using trim_string for emptiness, but sorts the original raw strings before selecting the first candidate, matching _path_group_from_finding instead of trimming before sort. Added test_plan_to_plan_reviewed_accepts_canonical_raw_write_scope_sorting to lock that behavior.

Validated locally with bash tests/test-validate-quest-state.sh: 52 passed.

- Reviewed by Codex, in collaboration with @KjellKod

File: .skills/quest/delegation/workflow.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/quest/delegation/workflow.md]: Fixed in 2e6019c. The workflow publish guard now validates the generated plan-phase backlog with validate-backlog --expected-phase plan before any atomic os.replace publish happens, so a review-phase backlog cannot replace the canonical artifact before the transition gate catches it.

Validated with bash scripts/quest_validate-manifest.sh and git diff --check -- .skills/quest/delegation/workflow.md.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_runtime/artifacts.py [hunk 14]
```diff
@@ -18,8 +18,7 @@
         "phase_01_plan",
         (
             "arbiter_verdict.md",
-            "review_findings.json",
-            "review_backlog.json",
+            "review_findings.json.next",
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py:21]: Fixed in af0d64a. Arbiter artifact preparation now preserves an existing arbiter_verdict.md instead of truncating it before each attempt, while still creating the verdict file when it is missing. That keeps the last canonical verdict intact if an arbiter retry crashes before producing replacement output.

Added artifact coverage for preserving an existing verdict and creating a missing one. Validated with python3 -m pytest -q, bash tests/test-quest-runtime.sh, bash tests/test-validate-quest-state.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_review_intelligence.py [hunk 15]
```diff
@@ -95,12 +96,31 @@ def _cmd_merge_findings(args: argparse.Namespace) -> int:
...
     print(json.dumps({"ok": True, "count": len(backlog["items"]), "output": args.output}, sort_keys=True))
     return 0
 
 
+def _cmd_validate_backlog(args: argparse.Namespace) -> int:
+    payload = _load_json(Path(args.input))
+    errors = validate_review_backlog(payload)
+    expected_phase = getattr(args, "expected_phase", None)
+    if expected_phase:
+        actual_phase = payload.get("phase") if isinstance(payload, dict) else None
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_review_intelligence.py:115]: Fixed in af0d64a. validate-backlog now has --strict-plan-defaults, which checks plan-phase items against the canonical build-backlog --phase plan defaults for decision, owner, and batch. The workflow pre-publish command now runs validate-backlog --expected-phase plan --strict-plan-defaults before os.replace, so drifted plan semantics cannot publish as canonical review_backlog.json.

Validated with python3 -m pytest -q, bash tests/test-quest-runtime.sh, bash tests/test-validate-quest-state.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_runtime/artifacts.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py:21]: Fixed in af0d64a. Artifact preparation now treats arbiter_verdict.md as preserve-on-prepare: it creates the file when missing, but no longer truncates an existing verdict before an arbiter retry. The arbiter continues to write arbiter_verdict.md on a successful attempt, while review_findings.json.next and review_backlog.json.next remain the scratch publish path.

Added artifact helper coverage that existing verdict content survives prepare_artifact_files(), and reran the runtime shell suite.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_review_intelligence.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_review_intelligence.py:115]: Fixed in af0d64a. validate-backlog now supports --strict-plan-defaults, which reuses the canonical Python plan-phase decision logic to reject drifted plan semantics before publish. The workflow now calls validate-backlog --expected-phase plan --strict-plan-defaults before atomic os.replace, and quest_validate-quest-state.sh uses the same strict CLI path instead of maintaining a separate jq copy.

Validated locally with:
- python3 -m pytest tests/unit/test_review_intelligence.py tests/unit/test_artifacts.py -q
- bash tests/test-validate-quest-state.sh
- bash tests/test-quest-runtime.sh
- bash scripts/quest_validate-manifest.sh
- git diff --check

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_runtime/artifacts.py [hunk 16]
```diff
@@ -49,6 +48,8 @@
     ),
 }
 
+PREPARE_PRESERVE_EXISTING_NAMES = frozenset({"arbiter_verdict.md"})
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py]: Fixed in df22a7e. The arbiter verdict now uses its own scratch path, arbiter_verdict.md.next, and both the arbiter instructions and runtime artifact map point there. The workflow publishes arbiter_verdict.md.next to canonical arbiter_verdict.md only after validation succeeds; failed validation keeps the prior canonical verdict untouched.

Updated the runtime retry harness to assert verdict, findings, and backlog canonical files remain unchanged until publish, and that the published verdict content matches the pre-publish .next artifact. Validated with python3 -m pytest -q, bash tests/test-quest-runtime.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: .skills/quest/delegation/workflow.md [hunk 17]
```diff
@@ -128,7 +128,7 @@ After any subagent completes, the orchestrator reads the agent's `handoff.json`
 5. **Artifact preparation (before every role invocation):**
    Before invoking any role, the orchestrator MUST:
    1. Resolve artifact paths: `expected_artifacts_for_role(quest_dir, phase, agent)`
-   2. Prepare files: `prepare_artifact_files(paths)` — creates parent directories and truncates files
+   2. Prepare files: `prepare_artifact_files(paths)` — creates parent directories and truncates role-output scratch files.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/quest/delegation/workflow.md]: Fixed in b15b5fc. cubic found the generic workflow wording no longer matched runtime behavior: prepare_artifact_files truncates every prepared path, not only scratch paths. The docs now state that explicitly and spell out the safety rule: retry-safe canonical files must stay out of expected_artifacts_for_role and use scratch paths plus validated publish instead.

Validated with bash tests/test-quest-runtime.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_runtime/artifacts.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py]: Fixed in df22a7e, with the artifact-preparation wording clarified in b15b5fc. Arbiter verdict output now uses arbiter_verdict.md.next, and canonical arbiter_verdict.md is only updated by the publish step after validation succeeds. The arbiter skill, workflow docs, runtime artifact list, and retry tests now all use the verdict scratch path.

Validated locally with:
- python3 -m pytest tests/unit/test_review_intelligence.py tests/unit/test_artifacts.py -q
- bash tests/test-quest-runtime.sh
- bash tests/test-validate-quest-state.sh
- bash scripts/quest_validate-manifest.sh
- git diff --check

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_allowlist_matcher.py [hunk 18]
```diff
@@ -58,14 +59,20 @@ def shell_tokens(command: str) -> list[str] | None:
         return None
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_allowlist_matcher.py:59]: Fixed in 2b4811e. token_prefix_matches() now normalizes the command executable for bare allowlist entries by basename, so safe absolute-path invocations like /usr/bin/find and /opt/homebrew/bin/rg match the same allowlist identity as find and rg. Path-pinned allowlist entries still require an exact first token. Added absolute-path allow tests for both find and rg in tests/unit/test_allowlist_matcher.py and reran python3 -m pytest -q tests/unit/test_allowlist_matcher.py, bash tests/test-quest-runtime.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/quest_installer.sh [hunk 19]
```diff
@@ -133,6 +133,16 @@ OLD_SCRIPT_NAMES=(
   "scripts/validate-quest-state.sh"
 )
 
+LEGACY_SOURCE_ONLY_TESTS=(
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_installer.sh]: Fixed in 2b4811e. The legacy cleanup list now only includes unit tests that Quest actually shipped in older installs: tests/unit/test_allowlist_matcher.py, tests/unit/test_codex_skill_wrappers.py, and tests/unit/test_review_intelligence.py. The newer repo-local source tests are no longer candidates for auto-prune, and the runtime shell harness now covers both pruning a previously shipped pristine file and preserving an unowned repo-local path like tests/unit/test_quest_complete.py. Validated with bash tests/test-quest-runtime.sh, bash scripts/quest_validate-manifest.sh, and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod

File: scripts/check_quest_checksum_drift.py [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+        line = raw.strip()
+        if not line or line.startswith("#"):
+            continue
+        try:
+            expected, relpath = raw.split("  ", 1)
+        except ValueError:
+            drift.append((raw, "malformed entry"))
+            continue
+
+        path = root / relpath
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/check_quest_checksum_drift.py]: Fixed in 69e042c. The drift helper now resolves each checksum path against the requested repo root and rejects any entry that escapes it before calling exists() or read_bytes(), so traversal or absolute paths are reported as unsafe instead of probing files outside the repo. Added focused unit coverage for both a matching in-repo file and an unsafe ../outside.txt entry, and validated with python3 -m pytest -q tests/unit/test_check_quest_checksum_drift.py and git diff --check.

- Reviewed by Codex, in collaboration with @KjellKod
